pkg/learning/mab.go

```diff
@@ -0,0 +1,158 @@
+	// for all choices.
+	avgReward := mab.sumRewards / float64(len(mab.choices))
+	if avgReward < -MABExponentThreshold || avgReward > MABExponentThreshold {
+		log.Logf(MABLogLevel, "MAB average reward (%v) exceeds threshold (+-%v), adjusting",
+			avgReward, MABExponentThreshold)
+		sumWeights := 0.0
+		for i := 0; i < len(mab.choices); i++ {
+			mab.choices[i].reward -= avgReward
+			mab.choices[i].weight = math.Exp(mab.choices[i].reward)
+			sumWeights += mab.choices[i].weight
+			mab.choices[i].sumWeights = sumWeights
+		}
+		mab.sumRewards = 0.0
+		mab.rewardAdjust += avgReward
+	}
+	// Update selection weight and prevent float64 overflow for individual
+	// choice.
+	weight := math.Exp(mab.choices[idx].reward)
+	if weight > MABWeightThresholdMax {
+		weight = MABWeightThresholdMax
+	}
+	if weight < MABWeightThresholdMin {
+		weight = MABWeightThresholdMin
+	}
+	mab.choices[idx].weight = weight
+	if idx == 0 {
+		mab.choices[idx].sumWeights = mab.choices[idx].weight
+	} else {
+		mab.choices[idx].sumWeights = mab.choices[idx-1].sumWeights + mab.choices[idx].weight
+	}
+	// Only need to update choices with a higher index.
+	for i := idx + 1; i < len(mab.choices); i++ {
+		mab.choices[i].sumWeights = mab.choices[i-1].sumWeights + mab.choices[i].weight
+	}
+}
+
+func (mab *MultiArmedBandit) NewChoiceWithReward(initialReward float64) int {
+	mab.mu.Lock()
+	defer mab.mu.Unlock()
+
+	newChoice := Choice{
+		reward:     initialReward - mab.rewardAdjust,
+		weight:     1.0,
+		sumWeights: 1.0,
+	}
+	if len(mab.choices) > 0 {
+		newChoice.sumWeights = mab.choices[len(mab.choices)-1].sumWeights + newChoice.weight
+	}
+	mab.choices = append(mab.choices, newChoice)
+	idx := len(mab.choices) - 1
+	mab.sumRewards += newChoice.reward
+	mab.updateWeight(idx)
+
+	// Need to update exploration and exploitation factors.
+	K := float64(len(mab.choices))
+	mab.eta = mab.theta * math.Sqrt(2.0*math.Log(K)/K)
+	mab.gamma = mab.eta / 2.0
+	log.Logf(MABLogLevel, "MAB gamma = %v, eta = %v", mab.gamma, mab.eta)
+
+	return idx
+}
+
+func (mab *MultiArmedBandit) Update(idx int, reward, pr float64) {
+	mab.mu.Lock()
+	defer mab.mu.Unlock()
+
+	if idx >= len(mab.choices) {
+		log.Fatalf("MAB Update Error: Index %v out of bound %v", idx, len(mab.choices))
+	}
+	if pr == 0 {
+		log.Fatalf("MAB Update Error: Probability is zero")
+	}
+	if reward > 1.0 || reward < -1.0 {
+		log.Fatalf("MAB Update Error: Reward (%v) should have been normalized to [-1, 1]", reward)
+	}
+	// Update reward based on growth factor.
+	rewardDiff := mab.eta * reward / (pr + mab.gamma)
+	log.Logf(MABLogLevel, "MAB reward update: %v * %v / (%v + %v) = %v",
+		mab.eta, reward, pr, mab.gamma, rewardDiff)
+	mab.choices[idx].reward += rewardDiff
+	mab.sumRewards += rewardDiff
+	mab.updateWeight(idx)
+}
+
+func (mab *MultiArmedBandit) GetRawReward(idx int) float64 {
+	mab.mu.Lock()
+	defer mab.mu.Unlock()
+
+	if idx < 0 || idx >= len(mab.choices) {
```

Comment:
This looks fishy. Did mask data races when mutex was omitted in some places?
Who is calling this with wrong idx and why?